packages/@aws-cdk/aws-codepipeline-api/lib/action.ts

```diff
@@ -0,0 +1,266 @@
+import events = require('@aws-cdk/aws-events');
+import iam = require('@aws-cdk/aws-iam');
+import cdk = require('@aws-cdk/cdk');
+import { Artifact } from './artifact';
+import validation = require('./validation');
+
+export enum ActionCategory {
+    Source = 'Source',
+    Build = 'Build',
+    Test = 'Test',
+    Approval = 'Approval',
+    Deploy = 'Deploy',
+    Invoke = 'Invoke'
+}
+
+/**
+ * Specifies the constraints on the number of input and output
+ * artifacts an action can have.
+ *
+ * The constraints for each action type are documented on the
+ * {@link https://docs.aws.amazon.com/codepipeline/latest/userguide/reference-pipeline-structure.html Pipeline Structure Reference} page.
+ */
+export interface ActionArtifactBounds {
+    minInputs: number;
+    maxInputs: number;
+    minOutputs: number;
+    maxOutputs: number;
+}
+
+export function defaultBounds(): ActionArtifactBounds {
+    return {
+        minInputs: 0,
+        maxInputs: 5,
+        minOutputs: 0,
+        maxOutputs: 5
+    };
+}
+
+export interface IBucketPermissions {
+    grantReadWrite(identity?: iam.IPrincipal): void;
+}
+
+export interface IStage {
+    readonly name: string;
+    readonly pipelineArn: cdk.Arn;
+    readonly pipelineRole: iam.Role;
+    readonly bucketPermissions: IBucketPermissions;
+    _addAction(action: Action): void;
+}
+
+export interface AbstractActionProps {
```

Comment:
The reason I did it this way is otherwise, we will copy & paste the same documentation for the `stage` property into 8+ different places. Granted, I didn't write any JsDocs for this property either, but that's a separate discussion ;p.

Reply:
Not sure that's a good enough reason... It's not such a complicated description...

Reply:
I actually have a second candidate property for this interface in mind: `runOrder`.

Does having 2 properties here make this interface valuable enough, in your opinion?